ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -288,34 +299,39 @@ boolean isChunkFillComplete() {
    * chunkFillerChannel, if there is any.
    * @throws InterruptedException if the call to get a chunk from the chunkFillerChannel is interrupted.
    */
-  void fillChunks()
-      throws InterruptedException {
-    PutChunk chunkToFill;
-    if (!chunkFillingCompleted && !operationCompleted) {
-      do {
-        // Attempt to fill a chunk
-        if (channelReadBuffer == null) {
-          channelReadBuffer = chunkFillerChannel.getNextChunk(0);
-        }
-        if (channelReadBuffer != null) {
-          chunkToFill = getChunkToFill();
-          if (chunkToFill == null) {
-            // no chunks are free to be filled yet.
-            break;
+  void fillChunks() {
+    try {
+      PutChunk chunkToFill;
+      if (!chunkFillingCompleted && !operationCompleted) {
+        do {
+          // Attempt to fill a chunk
+          if (channelReadBuffer == null) {
+            channelReadBuffer = chunkFillerChannel.getNextChunk(0);
           }
-          bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
-          if (!channelReadBuffer.hasRemaining()) {
-            chunkFillerChannel.resolveOldestChunk(null);
-            channelReadBuffer = null;
+          if (channelReadBuffer != null) {
+            chunkToFill = getChunkToFill();
+            if (chunkToFill == null) {
+              // no chunks are free to be filled yet.
+              break;
+            }
+            bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
+            if (!channelReadBuffer.hasRemaining()) {
+              chunkFillerChannel.resolveOldestChunk(null);
+              channelReadBuffer = null;
+            }
+          } else {
+            // channel does not have more data yet.
+            break;
           }
-        } else {
-          // channel does not have more data yet.
-          break;
+        } while (bytesFilledSoFar < blobSize);
+        if (bytesFilledSoFar == blobSize) {
+          chunkFillingCompleted = true;
         }
-      } while (bytesFilledSoFar < blobSize);
-      if (bytesFilledSoFar == blobSize) {
-        chunkFillingCompleted = true;
       }
+    } catch (Exception e) {
```

Comment:
To answer your question in the PR description - yes, if the chunk filling fails, we should simply complete the operation and invoke the callback, so what you do here is correct.

I was at first going to suggest that this `try-catch` be done in the chunk filler `run()` method, but I think there is a value in enclosing the `try-catch` here rather than outside - the value being the ability to set the operation exception and marking it as complete so that the other thread that polls can invoke the callback. Do check if there are similar reasons for the request response handler thread's poll - for why you chose to do it within the poll rather than outside - if there isn't, let us do it outside of the poll.